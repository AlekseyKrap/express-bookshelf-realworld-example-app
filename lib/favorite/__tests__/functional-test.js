'use strict';

const chance = require('chance').Chance('favorite');
const moment = require('moment');
const request = require('supertest');
const app = require('../../app');
const Article = require('../../article/model');
const Favorite = require('../model');
const User = require('../../user/model');

describe('favorite', () => {
  let user;
  let author;
  let article;

  beforeEach(async () => {
    user = await new User({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    }).save();
    author = await new User({
      bio: chance.profession(),
      email: chance.email(),
      image: chance.avatar(),
      password: chance.word({length: 8}),
      username: chance.word(),
    }).save();
    article = await new Article({
      author: author.id,
      body: chance.paragraph({sentences: 10}),
      description: chance.paragraph({sentences: 2}),
      title: chance.sentence({words: 3}),
    }).save();
    await new Favorite({
      user: author.id,
      article: article.id,
    }).save();
  });

  afterEach(async () => {
    await author.destroy();
    await user.destroy();
  });

  describe('POST /api/articles/:slug/favorite', () => {
    describe('without a logged in user', () => {
      test('should return a 401', async () => {
        const {statusCode, body} = await request(app)
          .post(`/api/articles/${article.get('slug')}/favorite`)
          .set('Accept', 'application/json')
          .send();

        expect(statusCode).toBe(401);
        expect(body).toMatchSnapshot();
      });
    });

    describe('with a logged in user', () => {
      describe('and an invalid slug', () => {
        test('should return a 404', async () => {
          const response = await request(app)
            .post('/api/articles/invalid-slug/favorite')
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${user.generateJWT()}`)
            .send();

          expect(response.statusCode).toBe(404);
          expect(response.body).toMatchSnapshot();
        });
      });

      describe('and a valid slug', () => {
        test('should return a 200', async () => {
          const response = await request(app)
            .post(`/api/articles/${article.get('slug')}/favorite`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${user.generateJWT()}`)
            .send();

          expect(response.statusCode).toBe(200);
          expect(response.body).toEqual({
            article: {
              author: {
                bio: author.get('bio'),
                following: false,
                image: author.get('image'),
                username: author.get('username'),
              },
              body: article.get('body'),
              createdAt: moment(article.get('created_at')).toISOString(),
              description: article.get('description'),
              favorited: true,
              favoritesCount: 2,
              slug: article.get('slug'),
              title: article.get('title'),
              updatedAt: moment(article.get('updated_at')).toISOString(),
            },
          });
        });
      });
    });
  });

  describe('DELETE /api/articles/:slug/favorite', () => {
    describe('without a logged in user', () => {
      test('should return a 401', async () => {
        const {statusCode, body} = await request(app)
          .delete(`/api/articles/${article.get('slug')}/favorite`)
          .set('Accept', 'application/json')
          .send();

        expect(statusCode).toBe(401);
        expect(body).toMatchSnapshot();
      });
    });

    describe('with a logged in user', () => {
      describe('and an invalid slug', () => {
        test('should return a 404', async () => {
          const response = await request(app)
            .delete('/api/articles/invalid-slug/favorite')
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${author.generateJWT()}`)
            .send();

          expect(response.statusCode).toBe(404);
          expect(response.body).toMatchSnapshot();
        });
      });

      describe('and a valid slug', () => {
        test('should return a 200', async () => {
          const response = await request(app)
            .delete(`/api/articles/${article.get('slug')}/favorite`)
            .set('Accept', 'application/json')
            .set('Authorization', `Token ${author.generateJWT()}`)
            .send();

          expect(response.statusCode).toBe(200);
          expect(response.body).toEqual({
            article: {
              author: {
                bio: author.get('bio'),
                following: false,
                image: author.get('image'),
                username: author.get('username'),
              },
              body: article.get('body'),
              createdAt: moment(article.get('created_at')).toISOString(),
              description: article.get('description'),
              favorited: false,
              favoritesCount: 0,
              slug: article.get('slug'),
              title: article.get('title'),
              updatedAt: moment(article.get('updated_at')).toISOString(),
            },
          });
        });
      });
    });
  });
});
