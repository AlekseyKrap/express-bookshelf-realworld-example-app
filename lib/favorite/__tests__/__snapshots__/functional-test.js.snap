// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favorite DELETE /api/articles/:slug/favorite with a logged in user and an invalid slug should return a 404 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "data": null,
      "isBoom": true,
      "isServer": false,
      "message": "Not Found",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Not Found",
          "message": "Not Found",
          "statusCode": 404,
        },
        "statusCode": 404,
      },
    },
    "message": "Not Found",
  },
}
`;

exports[`favorite DELETE /api/articles/:slug/favorite without a logged in user should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "credentials_required",
      "data": Object {
        "UnauthorizedError": Array [
          "No authorization token was found",
        ],
      },
      "inner": Object {
        "message": "No authorization token was found",
      },
      "isBoom": true,
      "isServer": false,
      "message": "No authorization token was found",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "No authorization token was found",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "No authorization token was found",
  },
}
`;

exports[`favorite POST /api/articles/:slug/favorite with a logged in user and an invalid slug should return a 404 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "data": null,
      "isBoom": true,
      "isServer": false,
      "message": "Not Found",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Not Found",
          "message": "Not Found",
          "statusCode": 404,
        },
        "statusCode": 404,
      },
    },
    "message": "Not Found",
  },
}
`;

exports[`favorite POST /api/articles/:slug/favorite without a logged in user should return a 401 1`] = `
Object {
  "errors": Object {
    "error": Object {
      "code": "credentials_required",
      "data": Object {
        "UnauthorizedError": Array [
          "No authorization token was found",
        ],
      },
      "inner": Object {
        "message": "No authorization token was found",
      },
      "isBoom": true,
      "isServer": false,
      "message": "No authorization token was found",
      "name": "UnauthorizedError",
      "output": Object {
        "headers": Object {},
        "payload": Object {
          "error": "Unauthorized",
          "message": "No authorization token was found",
          "statusCode": 401,
        },
        "statusCode": 401,
      },
      "status": 401,
    },
    "message": "No authorization token was found",
  },
}
`;
